core/peerset/src/slots.rs

```diff
@@ -0,0 +1,132 @@
+	/// Common peer is a type of peer that we stay connected to only if it's
+	/// useful for us.
+	Common,
+}
+
+/// Descibes why the reason of not being able to add given peer.
+pub enum SlotError {
+	/// Error returned when we are already connected to this peer.
+	AlreadyConnected(PeerId),
+	/// Error returned when max number of connections has been already established.
+	MaxConnections(PeerId),
+	/// Error returned when we should disconnect from a given common peer to make space
+	/// for a reserved peer.
+	DemandReroute {
+		/// Peer we should disconnect from.
+		disconnect: PeerId,
+		/// Peer we should connect to.
+		connect: PeerId,
+	}
+}
+
+/// Contains all information about group of slots.
+#[derive(Debug)]
+pub struct Slots {
+	max_slots: usize,
+	/// Nodes and their type. We use `LinkedHashMap` to make this data structure more predictable
+	slots: LinkedHashMap<PeerId, SlotType>,
+}
+
+impl Slots {
+	/// Creates a group of slots with a limited size.
+	pub fn new(max_slots: u32) -> Self {
+		let max_slots = max_slots as usize;
+		Slots {
+			max_slots,
+			slots: LinkedHashMap::with_capacity(max_slots),
+		}
+	}
+
+	/// Returns true if one of the slots contains given peer.
+	pub fn contains(&self, peer_id: &PeerId) -> bool {
+		self.slots.contains_key(peer_id)
+	}
+
+	/// Returns Ok if we successfully connected to a given peer.
+	pub fn add_peer(&mut self, peer_id: PeerId, slot_type: SlotType) -> Result<(), SlotError>  {
+		if let Some(st) = self.slots.get_mut(&peer_id) {
+			if let SlotType::Reserved = slot_type {
+				*st = SlotType::Reserved;
+			}
+			return Err(SlotError::AlreadyConnected(peer_id));
+		}
+
+		if self.slots.len() == self.max_slots {
+			if let SlotType::Reserved = slot_type {
+				if let Some((to_disconnect, _)) = self.slots.iter().find(|(_, &slot_type)| slot_type == SlotType::Common) {
+					return Err(SlotError::DemandReroute {
+						disconnect: to_disconnect.clone(),
+						connect: peer_id,
+					});
+				}
+			}
+			return Err(SlotError::MaxConnections(peer_id));
+		}
+
+		self.slots.insert(peer_id, slot_type);
+		Ok(())
+	}
+
+	pub fn clear_common_slots(&mut self) -> Vec<PeerId> {
+		let slots = mem::replace(&mut self.slots, LinkedHashMap::with_capacity(self.max_slots));
+		let mut common_peers = Vec::new();
+		for (peer_id, slot_type) in slots {
+			match slot_type {
+				SlotType::Common => {
+					common_peers.push(peer_id);
+				},
+				SlotType::Reserved => {
+					self.slots.insert(peer_id, slot_type);
+				},
+			}
+		}
+		common_peers
+	}
+
+	pub fn mark_reserved(&mut self, peer_id: &PeerId) {
+		if let Some(slot_type) = self.slots.get_mut(peer_id) {
+			*slot_type = SlotType::Reserved;
+		}
+	}
+
+	pub fn mark_not_reserved(&mut self, peer_id: &PeerId) {
+		if let Some(slot_type) = self.slots.get_mut(peer_id) {
+			*slot_type = SlotType::Common;
+		}
+	}
+
+	pub fn clear_slot(&mut self, peer_id: &PeerId) {
+		self.slots.remove(peer_id);
+	}
+
+	pub fn is_reserved(&self, peer_id: &PeerId) -> bool {
```

Comment:
Should be `is_connected_and_reserved` maybe?

Reply:
all peers in `slots` are connected

Reply:
Hence the name change to make that explicit.